keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,770 @@
+
+```
+type Container struct {
+     ......
+     // +optional
+     Lifecycle *Lifecycle `json:"lifecycle,omitempty" protobuf:"bytes,12,opt,name=lifecycle"`
+
+    // Notifiers can be triggered by Notification resources. What they mean are
+    // up to the Pod owner to define, but we may choose to define some commonly
+    // used names. Each notifier must have a unique name within the container
+    // definition, but the same name may be used in different containers of the
+    // same pod. When a Notification of a given name as a container notifier is created
+    // (and the pod to which the container belongs to is selected by the Notification's
+    // spec.podSelector), the notifier will run in the container.
+    // +optional
+    Notifiers []ContainerNotifier
+    ......  
+}
+
+type ContainerNotifier struct {
+   // Name must be unique within a container.
+   // Names are label-style, with an optional prefix. Names without the prefix
+   // such as `quiesce` are reserved for Kubernetes-defined "well-known" names.
+   // Names with a prefix such as `example.com/label-style` are custom names.
+   // This refers to a ContainerNotifierAction.
+   // Container notifier controller (or Kubelet in phase 2) executes the command
+   // defined in ContainerNotifierAction.
+   Name string
+
+   // ContainerNotifier must support idempotency. If the same command is run in 
+   // the same container multiple times, it should achieve the same results.
+   // For example, a quiesce command is run on a database and the database is
+   // quiesced. If the same quiesce command is run again, it should not have
+   // additional impact because the database is already quiesced.
+   Action *ContainerNotifierAction
+ }
+
+// ContainerNotifier describes a command to run in the container.
+type ContainerNotifierAction struct {
+    // handler describes how this notifier is delivered to the container.
+    Handler ContainerNotifierHandler
+
+    // Number of seconds after which the notifier times out.
+    // Defaults to 1 second. Minimum value is 1.
+    // +optional
+    TimeoutSeconds int32
+}
+
+// Handler defines a specific action that should be taken
+type ContainerNotifierHandler struct {
+    // Exec specifies the action to take.
+    // +optional
+    Exec *ExecAction `json:"exec,omitempty" protobuf:"bytes,1,opt,name=exec"`
+}
+```
+
+##### Notification API Object
+
+The Notification object is a request for a ContainerNotifier. It will be an in-tree API object in the "node.k8s.io" API group so that Kubelet can monitor and trigger the actions. The external controller or a user creates the Notification object to request a ContainerNotifier.
+
+All containers in a selected pod which have a "ContainerNotifier" defined in their spec with the "ContainerNotifierName" will have their corresponding "ContainerNotifierAction" executed. Moreover, there is no guarantee of ordering on execution.
+
+```
+type Notification struct {
```

Comment:
We need to define what happens in this case:

time 0: pods A, B, C are running, labelled foo=bar
time 1: create Notification selecting foo=bar
time 2: all notifiers in A, B, C are done
time 3: pod D is created, labelled foo=bar
*** Do we run the notifier on D?  If so, for how long does that apply?  If not, how do we prevent it?

Reply:
We have said that Kubelet (container notifier controller) will not do retries.  So if that is the case, Kubelet (container notifier controller) will only process the pods matching the label when the Notification object is first created?  If so, Pod D will not be processed?

Reply:
Is this a namespaced object ?  Once its created, it should apply to everything  until the api object is deleted . 
Is this more of an imperative api or we want it to be declarative (i guess the latter )?
i am thinking what does the declaration mean ?
- notify all containers that match this selector at this time t
- let there be notification for all pods matching this selector as long this api exists
- i want all containers of the pods matching a selector to be quiesced (this seems more natural)
its a bit confusing to think since the lifecycle hooks are tied to the lifecycle of the container/pod and are ok to be imperative, but this api is not tied to the lifecycle but more of a command to the containers.

Reply:
@krmayankk is on the issue.

This is a pseudo-imperative operation.

@xing-yang 


> Kubelet (container notifier controller) will only process the pods matching the label when the Notification object is first created

That seems reasonable at first blush, but what is the mechanism for that?  Is the rule "only applies to pods whose `metadata.creationTimestamp` is less than the Notification's `metadata.creationTimestamp` ?  If so, let's document it.  Also - less-than or less-than-or-equal?

Does this satisfy the higher-order controller's needs?  Is it plausible that a quiesce notification could be created, a snapshot started, then a new pod shows up which is not quiesced?  Even if Notifications applied to it, that would be async, so it still seems bad?

Reply:
@thockin using CreationTimestamp can certainly be one option to rule out newly created pods, it does not cover the case when a pod is evicted/died during hook execution. From quiescing requirements perspective, its critical to have a point-in-time "fixed" scope for snapshotting, thus if a pod died during the notification, we might want to fail the the whole notification.
we have a couple of other potential options here:
1. make notification to pod one-to-one by replacing podSelector with a pod definition. higher level controller like application snapshot controller then can have a pre-selected list and create one notification to each desired pod.
Advantage:

- simpler API
- ordering of notification is possible and more convenient to achieve for higher level controllers
disadvantages:
 a. probably does not fit well with the signal case where multiple pods need to be notified without a higher level controller

2. introducing a list of resolved pod in Notification.Spec which will be immutable once resolved. Pod selection is done prior to any real execution of notification. And that list would be the source of truth for the scope of execution.
WDYT?

Reply:
This is an interesting problem.  If a pod dies mid-snapshot, do you fail the whole snapshot?  If a new pod a starts mid-snapshot (scale-up?) does it need to quiesce?  Maybe that just "doesn't happen" ?

Your (1) happens to ALSO solve the status cardinality problem, so is attractive in that regard.  It means the scale of things kubelet has to watch just got much bigger, though, unless we set labels to scope it (e.g. set the node-name as a label, let kubelet watch only that).

But it is ugly for humans.

Another approach would be to define Notifications as applying to all pods.  In the example above it *would* apply to D.  And E, F, G etc until the Notification is removed.  All the controller (or kubelet) could guarantee would be that every Notification that selects a given pod would be attempted once, and if there were multiple notifications, they would be attempted in some deterministic order (e.g. creationTimestamp or RV).

So to expand the example:

t0:  pods A, B, C are running, labelled foo=bar
t1:  create Notification N1 selecting foo=bar
t2:  N1 notifiers in A, B, C are done
t3:  pod D is created, labelled foo=bar
t4:  N1 notifier in D is done
t5:  create Notification N2 selecting foo=bar
t6:  N2 notifiers in A, B, C, D are done
t7:  pod E is created, labelled foo=bar
t8:  N1 notifier in E is done
t9:  N2 notifier in E is done
t10: delete Notification N1
t11: pod F is created, labelled foo=bar
t12: N2 notifier in F is done  ## note N1 was not executed
t13: delete Notification N2

This seems maybe workable, but I need you to look at it from your use-cases.  Bear in mind that rapid sequences can be observed out of order by watchers, so if your plan would be to delete N1 then create N2 you need some time in between to know that all kubelets have observed the deletion (or ensure that it is "safe" to trigger N1 on new pods).  Similarly, if you want to clean up N1 and N2 you *must* delete them in order with sufficient time between to ensure no out-of-order deletion problems.

We don't have a "serializing barrier" operation. 

Regarding "ugly for humans", maybe we need both?  Individual pod-notifications (and status) which controllers can orchestrate PLUS a selector-notification which a built-in controller would expand into pod-notifications?  This represents a broader API change, but is the more flexible model.

Thoughts?

Regarding your (2) - does that satisfy requirements?  Let me see if I get it right:

t0:  pods A, B, C are running, labelled foo=bar
t1:  create Notification N1 selecting foo=bar
t2:  a controller evaluates foo=bar one time and notes the result (A, B, C)
t3:  N1 notifiers in A, B, C are done
t4:  pod D is created, labelled foo=bar
t5:  N1 notifier in D is NOT executed

Is that right?  How does that map to the use-cases?

Reply:
Applying to all pods for as long as the notification exists makes more sense from an api point of view. Also including the pod definition would be messy.

Reply:
> @krmayankk is on the issue.
> 
??

> This is a pseudo-imperative operation.
> 
Is that ok to have ? Do we need to update any apiconvention guidelines on when a pseudo imperative api makes more sense ?

Reply:
For the quiesce use case, a higher level backup controller needs to decide what pods are included. When a backup is started, the backup controller needs to figure out all the pods that need to get the quiesce command.  It can retrieve this information using a PodSelector.  After that it is a fixed list of Pods that need to be quiesced for this particular backup.  Therefore, we can’t have PodSelector in Notification.  It should be either a List of Pods that is immutable or a Pod name for a single pod, basically #2 and #1 suggested by Xiangqian.

If we decide to have a PodList in NotificationSpec, the steps would be like this:

1. An external Backup controller determines what pods need to receive the notification by using a PodSelector (this depends on backup controller logic).
2. Backup controller creates a Notification object with a PodList with the selected pods.

This way we have a concrete list of Pods. If quiesce fails for any pod/container, it should be treated as failure for the whole quiesce operation.  Typically the backup software gives user an option to complete a backup that is not quiesced successfully or fail it if any quiesce action fails. If user chooses to do a backup even if quiesce fails, the backup controller will trigger a un-quiesce command to undo anything that is already quiesced but skip doing quiesce for the remaining pod/container.  Then the backup controller will take a snapshot and mark that snapshot as not-quiesced.  User can decide what to do with a snapshot that is not quiesced.

Reply:
> we can’t have PodSelector in Notification

I'm not sure I buy that.  This API needs to be general enough to use by humans.  We need to be clear about what happens if a new pod is created that satisfies the criteria.  Don't over-focus on just quiesce